# HomeWork 2




Question2: KNN Classification in sklearn
Write and submit your python codes in “Jupyter Notebook” to perform the following tasks
(submit the .ipynb file). Make sure to provide proper descriptions as MarkDown for each section
of your code (each section of the code must have a short meaningful description right before
that, describing what this part of the code is supposed to do!)

---------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------


### (A) Read the	iris	dataset	from	the	following	URL: 'https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv'and	assign	it	to a	Pandas	DataFrame	as	you	learned in	tutorial	Lab2-3.

In [56]:
# The following line will import KNeighborsClassifier "Class"
# KNeighborsClassifier is name of a "sklearn class" to perform "KNN Classification" 
from sklearn.neighbors import KNeighborsClassifier

# Importing the required packages and libraries
# we will need numpy and pandas later
import numpy as np
import pandas as pd

iris_df = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')

from sklearn.datasets import load_iris 

iris = load_iris()

X = iris.data  # X will be feature matrix
print(X)



[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.  3.2 1.2 0.2]
 [5.5 3.5 1.3 0.2]
 [4.9 3.1 1.5 0.1]
 [4.4 3.  1.3 0.2]
 [5.1 3.4 1.5 0.2]
 [5.  3.5 1.3 0.3]
 [4.5 2.3 1.3 0.3]
 [4.4 3.2 1.3 0.2]
 [5.  3.5 1.6 0.6]
 [5.1 3.8 1.9 0.4]
 [4.8 3.  1.4 0.3]
 [5.1 3.8 1.6 0.2]
 [4.6 3.2 1.4 0.2]
 [5.3 3.7 1.5 0.2]
 [5.  3.3 1.4 0.2]
 [7.  3.2 4.7 1.4]
 [6.4 3.2 4.5 1.5]
 [6.9 3.1 4.

In [13]:
iris_df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
5             5.4          3.9           1.7          0.4     setosa
6             4.6          3.4           1.4          0.3     setosa
7             5.0          3.4           1.5          0.2     setosa
8             4.4          2.9           1.4          0.2     setosa
9             4.9          3.1           1.5          0.1     setosa
10            5.4          3.7           1.5          0.2     setosa
11            4.8          3.4           1.6          0.2     setosa
12            4.8          3.0           1.4          0.1     setosa
13            4.3          3.0           1.1          0.1     setosa
14            5.8          4.0           1.2          0.2     setosa
15            5.7          4.4           1.5          0.4     setosa
16            5.4          3.9           1.3          0.4     setosa
17            5.1          3.5           1.4          0.3     setosa
18            5.7          3.8           1.7          0.3     setosa
19            5.1          3.8           1.5          0.3     setosa
20            5.4          3.4           1.7          0.2     setosa
21            5.1          3.7           1.5          0.4     setosa
22            4.6          3.6           1.0          0.2     setosa
23            5.1          3.3           1.7          0.5     setosa
24            4.8          3.4           1.9          0.2     setosa
25            5.0          3.0           1.6          0.2     setosa
26            5.0          3.4           1.6          0.4     setosa
27            5.2          3.5           1.5          0.2     setosa
28            5.2          3.4           1.4          0.2     setosa
29            4.7          3.2           1.6          0.2     setosa
..            ...          ...           ...          ...        ...
120           6.9          3.2           5.7          2.3  virginica
121           5.6          2.8           4.9          2.0  virginica
122           7.7          2.8           6.7          2.0  virginica
123           6.3          2.7           4.9          1.8  virginica
124           6.7          3.3           5.7          2.1  virginica
125           7.2          3.2           6.0          1.8  virginica
126           6.2          2.8           4.8          1.8  virginica
127           6.1          3.0           4.9          1.8  virginica
128           6.4          2.8           5.6          2.1  virginica
129           7.2          3.0           5.8          1.6  virginica
130           7.4          2.8           6.1          1.9  virginica
131           7.9          3.8           6.4          2.0  virginica
132           6.4          2.8           5.6          2.2  virginica
133           6.3          2.8           5.1          1.5  virginica
134           6.1          2.6           5.6          1.4  virginica
135           7.7          3.0           6.1          2.3  virginica
136           6.3          3.4           5.6          2.4  virginica
137           6.4          3.1           5.5          1.8  virginica
138           6.0          3.0           4.8          1.8  virginica
139           6.9          3.1           5.4          2.1  virginica
140           6.7          3.1           5.6          2.4  virginica
141           6.9          3.1           5.1          2.3  virginica
142           5.8          2.7           5.1          1.9  virginica
143           6.8          3.2           5.9          2.3  virginica
144           6.7          3.3           5.7          2.5  virginica
145           6.7          3.0           5.2          2.3  virginica
146

# b- Split the dataset into testing and training sets with the following parameters: test_size=0.4, random_state=6

In [14]:
# "feature_names" attribute will return the name of features:
print(iris.feature_names)

['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


In [15]:
print(X.shape) # this line print the size of iris.data (iris feature matrix)

(150, 4)


In [25]:
# "target" attribute will return the iris dataset labels 
# for the sklearn embedded iris dataset, the labels are already converted to numeric
y = iris.target  # y will be label vector
print(y)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


In [26]:
print(y.shape) # this line print the size of iris.target

(150,)


In [27]:
# "target_names" attribute will return the name of encoded labels: 0 = setosa, 1 = versicolor, 2 = virginica
print(iris.target_names)

['setosa' 'versicolor' 'virginica']


In [28]:
# In the following line, "knn" is instantiated as an "object" of KNeighborsClassifier "class". 
k = 1
knn = KNeighborsClassifier(n_neighbors=k) 

# We can use the method "fit" of the "object knn" along with training dataset and labels to train the model.
knn.fit(X, y)

# We can use the method "predict" of the *trained* object knn on one or more testing data sample to perform prediction:

X_Testing = [[6, 3, 5.9, 2.9]]

y_predict = knn.predict(X_Testing)

print(y_predict)


[2]


In [29]:
# We can use the method "predict" of the *trained* object knn on one or more testing data sample to perform prediction:
# Two new data samples:
X_Testing = [[6, 3, 5.9, 2.9],[3.2, 3, 1.9, 0.3]]

y_predict = knn.predict(X_Testing)

print(y_predict)

[2 0]


In [30]:
# Defining a function to convert "categorical" labels to "numerical" labels
# This is optional, because the latest revision of sklearn accepts non-numerical labels too!

def categorical_to_numeric(x):
    if x == 'setosa':
        return 0
    elif x == 'versicolor':
        return 1
    elif x == 'virginica':
        return 2

In [31]:
# Applying the function on species column and adding corrsponding numerical label column:
iris_df['label'] = iris_df['species'].apply(categorical_to_numeric)

# checking the dataset by printing every 10 lines:

iris_df[0::10]


# Creating the Feature Matrix for iris dataset:

# create a python list of feature names that would like to pick from the dataset:
feature_cols = ['sepal_length','sepal_width','petal_length','petal_width']

# use the above list to select the features from the original DataFrame
X = iris_df[feature_cols]  

# print the first 5 rows
X



sepal_length  sepal_width  petal_length  petal_width
0             5.1          3.5           1.4          0.2
1             4.9          3.0           1.4          0.2
2             4.7          3.2           1.3          0.2
3             4.6          3.1           1.5          0.2
4             5.0          3.6           1.4          0.2
5             5.4          3.9           1.7          0.4
6             4.6          3.4           1.4          0.3
7             5.0          3.4           1.5          0.2
8             4.4          2.9           1.4          0.2
9             4.9          3.1           1.5          0.1
10            5.4          3.7           1.5          0.2
11            4.8          3.4           1.6          0.2
12            4.8          3.0           1.4          0.1
13            4.3          3.0           1.1          0.1
14            5.8          4.0           1.2          0.2
15            5.7          4.4           1.5          0.4
16            5.4          3.9           1.3          0.4
17            5.1          3.5           1.4          0.3
18            5.7          3.8           1.7          0.3
19            5.1          3.8           1.5          0.3
20            5.4          3.4           1.7          0.2
21            5.1          3.7           1.5          0.4
22            4.6          3.6           1.0          0.2
23            5.1          3.3           1.7          0.5
24            4.8          3.4           1.9          0.2
25            5.0          3.0           1.6          0.2
26            5.0          3.4           1.6          0.4
27            5.2          3.5           1.5          0.2
28            5.2          3.4           1.4          0.2
29            4.7          3.2           1.6          0.2
..            ...          ...           ...          ...
120           6.9          3.2           5.7          2.3
121           5.6          2.8           4.9          2.0
122           7.7          2.8           6.7          2.0
123           6.3          2.7           4.9          1.8
124           6.7          3.3           5.7          2.1
125           7.2          3.2           6.0          1.8
126           6.2          2.8           4.8          1.8
127           6.1          3.0           4.9          1.8
128           6.4          2.8           5.6          2.1
129           7.2          3.0           5.8          1.6
130           7.4          2.8           6.1          1.9
131           7.9          3.8           6.4          2.0
132           6.4          2.8           5.6          2.2
133           6.3          2.8           5.1          1.5
134           6.1          2.6           5.6          1.4
135           7.7          3.0           6.1          2.3
136           6.3          3.4           5.6          2.4
137           6.4          3.1           5.5          1.8
138           6.0          3.0           4.8          1.8
139           6.9          3.1           5.4          2.1
140           6.7          3.1           5.6          2.4
141           6.9          3.1           5.1          2.3
142           5.8          2.7           5.1          1.9
143           6.8          3.2           5.9          2.3
144           6.7          3.3           5.7          2.5
145           6.7          3.0           5.2          2.3
146           6.3          2.5           5.0          1.9
147           6.5          3.0           5.2          2.0
148           6.2          3.4           5.4          2.3
149           5.9          3.0           5.1          1.8

[150 rows x 4 columns]

In [32]:
# checking the size of Feature Matix X:

print(X.shape)

(150, 4)


In [33]:
# select a Series of labels (the last column) from the DataFrame
# y = iris_df['label'] # this is the index that we gave to the labels
# OR:
y = iris_df['species'] # this is the original categorical labels (the latest revision of sklearn accepts non-numerical labels)

# checking the label vector by printing every 10 values
y[::10]

0          setosa
10         setosa
20         setosa
30         setosa
40         setosa
50     versicolor
60     versicolor
70     versicolor
80     versicolor
90     versicolor
100     virginica
110     virginica
120     virginica
130     virginica
140     virginica
Name: species, dtype: object

In [34]:
# Instantiating another "object" of KNeighborsClassifier "class" with k=3:
k = 3
my_knn_for_cs4661 = KNeighborsClassifier(n_neighbors=k) # name of the object is arbitrary!

# We use the method "fit" of the object along with training dataset and labels to train the model.
my_knn_for_cs4661.fit(X, y)

X_Testing = [[6, 3, 5.9, 2.9],[3.2, 3, 1.9, 0.3]]
y_predict = my_knn_for_cs4661.predict(X_Testing)
print(y_predict)
print('\n')

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

# print the size of the traning set:
print(X_train.shape)
print(y_train.shape)
print('\n')


# print the size of the testing set:
print(X_test.shape)
print(y_test.shape)
print('\n')



['virginica' 'setosa']


(90, 4)
(90,)


(60, 4)
(60,)




In [35]:
print(X_test)

     sepal_length  sepal_width  petal_length  petal_width
4             5.0          3.6           1.4          0.2
116           6.5          3.0           5.5          1.8
2             4.7          3.2           1.3          0.2
23            5.1          3.3           1.7          0.5
123           6.3          2.7           4.9          1.8
96            5.7          2.9           4.2          1.3
134           6.1          2.6           5.6          1.4
39            5.1          3.4           1.5          0.2
137           6.4          3.1           5.5          1.8
53            5.5          2.3           4.0          1.3
127           6.1          3.0           4.9          1.8
81            5.5          2.4           3.7          1.0
115           6.4          3.2           5.3          2.3
135           7.7          3.0           6.1          2.3
74            6.4          2.9           4.3          1.3
119           6.0          2.2           5.0          1.5
105           

In [45]:
print(y_test)

4          setosa
116     virginica
2          setosa
23         setosa
123     virginica
96     versicolor
134     virginica
39         setosa
137     virginica
53     versicolor
127     virginica
81     versicolor
115     virginica
135     virginica
74     versicolor
119     virginica
105     virginica
51     versicolor
92     versicolor
32         setosa
37         setosa
120     virginica
44         setosa
0          setosa
55     versicolor
72     versicolor
87     versicolor
102     virginica
30         setosa
93     versicolor
45         setosa
59     versicolor
16         setosa
13         setosa
133     virginica
128     virginica
64     versicolor
146     virginica
95     versicolor
49         setosa
17         setosa
103     virginica
71     versicolor
61     versicolor
46         setosa
12         setosa
52     versicolor
27         setosa
34         setosa
54     versicolor
118     virginica
117     virginica
121     virginica
6          setosa
111     virginica
18        

# c- Instantiate a KNN object with K=3, train it on the training set and test it on the testing set.
#  Then, calculate the accuracy of your prediction as you learned in Lab3.

In [47]:
# Training ONLY on the training set:
my_knn_for_cs4661.fit(X_train, y_train)

# Testing on the testing set:
y_predict = my_knn_for_cs4661.predict(X_test)
print(y_predict)

['setosa' 'virginica' 'setosa' 'setosa' 'versicolor' 'versicolor'
 'virginica' 'setosa' 'virginica' 'versicolor' 'versicolor' 'versicolor'
 'virginica' 'virginica' 'versicolor' 'versicolor' 'virginica'
 'versicolor' 'versicolor' 'versicolor' 'setosa' 'virginica' 'setosa'
 'setosa' 'versicolor' 'versicolor' 'versicolor' 'virginica' 'setosa'
 'versicolor' 'setosa' 'versicolor' 'versicolor' 'setosa' 'versicolor'
 'virginica' 'versicolor' 'versicolor' 'versicolor' 'setosa' 'setosa'
 'virginica' 'versicolor' 'versicolor' 'setosa' 'setosa' 'virginica'
 'setosa' 'setosa' 'versicolor' 'virginica' 'virginica' 'versicolor'
 'setosa' 'virginica' 'versicolor' 'setosa' 'versicolor' 'versicolor'
 'virginica']


In [48]:

from sklearn.metrics import accuracy_score

# Example:
y_pred    = [0, 2, 1, 1]
y_actual  = [0, 1, 2, 1]

score = accuracy_score(y_actual, y_pred)

print(score)

0.5


In [50]:

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.8166666666666667


#  d- Repeat part (c) for K=1, K=5, K=7, K=11, K=15, K=27, K=59 (you can simply use a “for loop,” and save the final accuracy results in a list). Does the accuracy always get better by increasing the value K?

In [53]:
listK = [1,5,7,11,13,15,27,59,]
ScoreK = []
AccuracyK = []

for k in listK:
    my_knn_for_cs4661 = KNeighborsClassifier(n_neighbors=k) # name of the object is arbitrary!
    
    my_knn_for_cs4661.fit(X, y)


    X_Testing = [[6, 3, 5.9, 2.9],[3.2, 3, 1.9, 0.3]]
    y_predict = my_knn_for_cs4661.predict(X_Testing)

    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

    my_knn_for_cs4661.fit(X_train, y_train)


    y_predict = my_knn_for_cs4661.predict(X_test)
    
    from sklearn.metrics import accuracy_score
    
    y_pred    = [0, 2, 1, 1]
    y_actual  = [0, 1, 2, 1]
    score = accuracy_score(y_actual, y_pred)
    
    ScoreK.append(score)

    from sklearn.metrics import accuracy_score
    accuracy = accuracy_score(y_test, y_predict)

    AccuracyK.append(accuracy)
    
print(listK)
print('\n')
print(ScoreK)
print('\n')
print(AccuracyK)


[1, 5, 7, 11, 13, 15, 27, 59]


[0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]


[0.95, 0.9833333333333333, 0.9666666666666667, 0.9666666666666667, 0.9666666666666667, 0.9333333333333333, 0.9166666666666666, 0.8166666666666667]


# Does the accuracy always get better by increasing K?
 
## Answer: No

# e- Now, suppose that we would like to make prediction based on only ONE single feature.To find the best single feature, we have to try every feature individually. In other word,we have to repeat part (c) with K=3, four times (each time using only one of the 4 features), and compute the accuracy each time. Then, check the accuracies.Which individual feature provide the best accuracy (the best feature)? Which one is the second best feature? (Note: you have to train, test, and evaluate your model 4 times, each time on a dataset including only one of the features, and save the final accuracy results in a list).

In [54]:
k=3

knn_hw2_e= KNeighborsClassifier(n_neighbors=k)

knn_result_dict={}

for f in feature_cols:
    
        X_feature=iris_df[f]

        X_ftrain,X_ftest,y_ftrain,y_ftest = train_test_split (X_feature.values.reshape(len(X_feature),1),y,test_size=0.4,random_state=1)

        knn_hw2_e.fit(X_ftrain,y_ftrain)

        y_fpredict=knn_hw2_e.predict(X_ftest)

        accuracy = accuracy_score(y_ftest,y_fpredict)

        knn_result_dict[f]=accuracy
        
    
print(knn_result_dict)


{'sepal_length': 0.6333333333333333, 'sepal_width': 0.5333333333333333, 'petal_length': 0.9666666666666667, 'petal_width': 0.9333333333333333}


# Which Individual features provides the best accuracy(the best feature)?
### Answer: petal_length with 0.966666666667 accuracy
## Which one is the second best features ?
### Answer: petal_width with 0.93333333333333335  accuracy

#   f- Now, we want to repeat part (e), this time using two features. you have to train, test, and evaluate your model for 6 different cases: using (1st and 2nd features), (1st and 3rd features), (1st and 4th features), (2nd and 3rd features), (2nd and 4th features), (3rd and 4th features)! Which “feature pair” provides the best accuracy?

In [55]:
# we perform the KNeighborsClassifier on different feature pairs and store each one's accuracy in knn_result 

k=3

knn_hw2_e= KNeighborsClassifier(n_neighbors=k)

features =[['sepal_length','sepal_width'],['sepal_length','petal_length'],['sepal_length','petal_width'],['sepal_width','petal_length'],['sepal_width','petal_width'],['petal_length','petal_width']]

knn_result={}

for f in features:
    
        X_feature=iris_df[f]

        X_ftrain,X_ftest,y_ftrain,y_ftest = train_test_split (X_feature,y,test_size=0.4,random_state=1)

        knn_hw2_e.fit(X_ftrain,y_ftrain)

        y_fpredict=knn_hw2_e.predict(X_ftest)

        accuracy = accuracy_score(y_ftest,y_fpredict)

        knn_result[','.join(f)]=accuracy

print(knn_result)



for k in knn_result.keys():
        
        print(k,":",knn_result[k])
    

{'sepal_length,sepal_width': 0.65, 'sepal_length,petal_length': 0.9333333333333333, 'sepal_length,petal_width': 0.9833333333333333, 'sepal_width,petal_length': 0.95, 'sepal_width,petal_width': 0.9666666666666667, 'petal_length,petal_width': 0.9666666666666667}
sepal_length,sepal_width : 0.65
sepal_length,petal_length : 0.9333333333333333
sepal_length,petal_width : 0.9833333333333333
sepal_width,petal_length : 0.95
sepal_width,petal_width : 0.9666666666666667
petal_length,petal_width : 0.9666666666666667


# Which “feature	pair” provide	the	best	accuracy?

## Ans: feature pair ('sepal_length','petal_width') with 0.983333333333 accuracy

### (G) BigQuestion: Does the	“best	feature	pair” from	part	(f) contain of both “first best	feature”and	“second	best	feature” from	part	(e)?	 In	other	word,	can	we	conclude	that	the	“best	two	features”	for	classification	are	the	“first	best	individual	feature”	and	the	“second	best	individual	feature”	together?


#### Ans: No, "best features pair" (sepal_length, petal_width) with accuracy 0.983333333333 does not contain the  first best feature "petal_length" and second best feature "petal_width", So best feature pair is more accurate than first and second best  individual features, we can say that "best two features" for classification are not “first	best	individual	feature”	and	the	“second	best	individual	feature”	together.

### (H) Optional	Question:	Justify	your	answer	to	part	(g)!		If	yes,	why?		If	no,	why	not?

#### Ans. When we take one feature at one time then we consider only that feature an it's distance from testing data. while we take pair of features then we have to consider both feature and it's distance from testing sample .so, everytime it is not necessary that first and second individual feature together give a best accuracy.
